### PR TITLE
layer fusion patches for dnn module

### PR DESCRIPTION
**Pull request by pengli**
```
allow_multiple_commits=1
```

### CODE REVIEW

File: modules/dnn/include/opencv2/dnn/all_layers.hpp [hunk 1]
```diff
@@ -473,6 +473,7 @@ CV__DNN_EXPERIMENTAL_NS_BEGIN
         float epsilon;
 
         virtual void getScaleShift(Mat& scale, Mat& shift) const = 0;
+        virtual void getScaleShift(UMat& scale, UMat& shift) const = 0;
```
Comment (alalek) [inline modules/dnn/include/opencv2/dnn/all_layers.hpp]: This call is not necessary. It is used during preparation stage, so there is no problem to use Mat instead.

Also
```
         umat_weight = weights_.getUMat(ACCESS_READ);
         umat_bias = bias_.getUMat(ACCESS_READ);
```

in batch_norm_layer.cpp should not be initialized unconditionally.

On systems with OpenCL without "zero-copy" support memory consumption would be too large (especially when OpenCL is not used).
It is better to initialize this due the first call of the "forward()" processing.

File: modules/dnn/src/dnn.cpp [hunk 2]
```diff
@@ -1245,6 +1246,7 @@ struct Net::Impl
                          nextData &&
                         (!nextData->type.compare("ReLU") ||
                          !nextData->type.compare("ChannelsPReLU") ||
+                         !nextData->type.compare("TanH") ||
```
Comment (alalek) [inline modules/dnn/src/dnn.cpp]: I believe this form `0 == nextData->type.compare("...")` looks more clear than negative `compare()` call

Please update the comment above too.

Comment (pengli) [reply modules/dnn/src/dnn.cpp]: yes,  I changed to "==" operator to make it clear

File: modules/dnn/include/opencv2/dnn/all_layers.hpp (hunk 1 above)
Comment (pengli) [reply modules/dnn/include/opencv2/dnn/all_layers.hpp]: it makes sense to me.  code is updated accordingly

**Pull request merged**